### PR TITLE
Use a different map generator configuration for each feature

### PR DESCRIPTION
**Pull request by jaimerson**
### Usage
```ruby
# features/my_shiny_feature_spec.rb
require 'feature_helper'

RSpec.describe 'My feature with map generator' do
  include_context 'simple git repository'
  let(:map_generator_config) do
    # must return a string containing the code that will be written in spec_helper.rb
    <<~CONFIG
      Crystalball::MapGenerator.start! do |c|
        c.register Crystalball::MapGenerator::CoverageStrategy.new
      end
      Crystalball::Rails::TablesMapGenerator.start! # if it also needs a table map
    CONFIG
  end
  # ...
end
```

### CODE REVIEW

File: features/evaluation_spec.rb [hunk 1]
```diff
@@ -9,6 +9,21 @@
   end
   include_context 'simple git repository'
 
+  let(:map_generator_config) do
+    <<~CONFIG
+      Crystalball::MapGenerator.start! do |c|
```
Comment (EPecherkin) [inline features/evaluation_spec.rb]: I think we should use CoverageStrategy only for this feature

File: features/evaluation_spec.rb [hunk 2]
```diff
@@ -27,6 +42,7 @@
       class Model1
       end
     RUBY
+    git.checkout(spec_helper)
```
Comment (EPecherkin) [inline features/evaluation_spec.rb]: Let's commit this changes before we run anything. So, we will be on a safe side forever

File: features/modified_execution_paths/change_in_branch_spec.rb [hunk 3]
```diff
@@ -8,6 +8,16 @@
   include_context 'model1 examples'
   include_context 'base forecast'
 
+  let(:map_generator_config) do
+    <<~CONFIG
+      Crystalball::MapGenerator.start! do |c|
+        c.register Crystalball::MapGenerator::CoverageStrategy.new
+        c.register Crystalball::MapGenerator::DescribedClassStrategy.new
+        c.register Crystalball::MapGenerator::AllocatedObjectsStrategy.build(only: ['Object'])
```
Comment (EPecherkin) [inline features/modified_execution_paths/change_in_branch_spec.rb]: I think it is not necessary to have all of these here. We can stay with `CoverageStrategy` only

File: features/modified_execution_paths/change_spec.rb [hunk 4]
```diff
@@ -8,6 +8,16 @@
   include_context 'model1 examples'
   include_context 'base forecast'
 
+  let(:map_generator_config) do
+    <<~CONFIG
+      Crystalball::MapGenerator.start! do |c|
+        c.register Crystalball::MapGenerator::CoverageStrategy.new
+        c.register Crystalball::MapGenerator::AllocatedObjectsStrategy.build(only: ['Object'])
```
Comment (EPecherkin) [inline features/modified_execution_paths/change_spec.rb]: and here as well. we can go with `CoverageStrategy` only

File: features/modified_schema/change_spec.rb [hunk 5]
```diff
@@ -2,10 +2,23 @@
...
+RSpec.describe 'Changing schema file' do
   include_context 'simple git repository'
   include_context 'model1 examples'
   include_context 'base forecast'
+  let(:schema_path) { root.join('db', 'schema.rb') }
+
+  let(:map_generator_config) do
+    <<~CONFIG
+      Crystalball::MapGenerator.start! do |c|
+        c.register Crystalball::MapGenerator::CoverageStrategy.new
```
Comment (EPecherkin) [inline features/modified_schema/change_spec.rb]: `TablesMapGenerator` + `CoverageStrategy` should be enough I think

File: features/rspec_runner_spec.rb [hunk 6]
```diff
@@ -10,6 +10,21 @@
   let(:important_class_path) { root.join('lib/important_class.rb') }
   let(:other_important_class_path) { root.join('lib/other_important_class.rb') }
 
+  let(:map_generator_config) do
+    <<~CONFIG
+      Crystalball::MapGenerator.start! do |c|
+        c.register Crystalball::MapGenerator::CoverageStrategy.new
+        c.register Crystalball::MapGenerator::AllocatedObjectsStrategy.build(only: ['Object'])
```
Comment (EPecherkin) [inline features/rspec_runner_spec.rb]: We can drop everything except `CoverageStrategy` as well I think

File: features/support/shared_contexts/simple_git_repository.rb [hunk 7]
```diff
@@ -8,24 +8,21 @@
...
   let(:spec_path) { root.join('spec') }
   let(:class1_spec_path) { spec_path.join('class1_spec.rb') }
-  let(:factories_path) { spec_path.join('factories') }
-  let(:model1_factory_path) { factories_path.join('model1s.rb') }
-  let(:model1_factory_modification_path) { factories_path.join('model1s_modification.rb') }
   let(:action_view_shared_context) { spec_path.join('support', 'shared_contexts', 'action_view.rb') }
   let(:git) { Git.init(root.to_s) }
+  let(:spec_helper) { File.join(root, 'spec/spec_helper.rb') }
+
+  # By default, generate map just with metadata
```
Comment (EPecherkin) [inline features/support/shared_contexts/simple_git_repository.rb]: `generate map with metadata only` sound a bit better I think. but it is a stylish note, you can ignore it if you like original variant more :)

File: features/support/shared_contexts/simple_git_repository.rb [hunk 8]
```diff
@@ -34,13 +31,19 @@
...
+    raise "Can't generate map" unless generate_map
   end
 
   after do
     root.rmtree
   end
 
+  def generate_map
+    config = map_generator_config.to_s.chomp
+    replace(spec_helper, /# MAP_GENERATOR_CONFIG/, config)
```
Comment (EPecherkin) [inline features/support/shared_contexts/simple_git_repository.rb]: here you change `spec_helper` and generate a map
let's split this logic
it is better to change `spec_helper` and commit it separately. so, main logic of a feature will be executed on a clean repo without unexpected side-effects

**Review by EPecherkin — DISMISSED** [review]
1. commit changed `spec_helper`, so we will have a clean repo at the beginning of each feature
2. use `CoverageStrategy` only unless you want to test another strategy. it will make features more granular

File: features/support/shared_contexts/simple_git_repository.rb (hunk 7 above)
Comment (jaimerson) [reply features/support/shared_contexts/simple_git_repository.rb]: You won't believe it, but I wrote exactly that and changed to this :smile:
I'm fine with either, so I'll change it.

File: features/evaluation_spec.rb (hunk 1 above)
Comment (jaimerson) [reply features/evaluation_spec.rb]: Done

File: features/modified_execution_paths/change_spec.rb [hunk 9]
```diff
@@ -8,6 +8,15 @@
   include_context 'model1 examples'
   include_context 'base forecast'
 
+  let(:map_generator_config) do
+    <<~CONFIG
+      Crystalball::MapGenerator.start! do |c|
+        c.register Crystalball::MapGenerator::AllocatedObjectsStrategy.build(only: ['Object'])
+        c.register Crystalball::MapGenerator::DescribedClassStrategy.new
```
Comment (jaimerson) [inline features/modified_execution_paths/change_spec.rb]: Here we need `AllocatedObjects` to detect `Class2.new` in `file_spec.rb` and `DescribedClass` for the `class1` examples.

Comment (pluff) [inline features/modified_execution_paths/change_spec.rb]: Lets split it into 2 different features one for each strategy

File: features/evaluation_spec.rb [hunk 10]
```diff
@@ -9,6 +9,14 @@
   end
   include_context 'simple git repository'
 
+  let(:map_generator_config) do
```
Comment (pluff) [inline features/evaluation_spec.rb]:
```
map_generator_config do
  "STRING"
end
```
will be a bit more declarative

File: features/modified_execution_paths/change_spec.rb [hunk 11]
```diff
@@ -84,49 +96,87 @@
...
-                                      './spec/models/model1_spec.rb[1:4:1]'
-                                    ])
+  context 'when dealing with factories' do
+    let(:map_generator_config) do
+      <<~CONFIG
+        Crystalball::MapGenerator.start! do |c|
+          c.register Crystalball::MapGenerator::FactoryBotStrategy.new
+        end
+      CONFIG
+    end
```
Comment (pluff) [inline features/modified_execution_paths/change_spec.rb]: lets split features for `ModifiedExecutuonPaths` prediction strategy by execution strategy too.
 E.g. `features/modified_execution_paths/factory_boty_strategy_spec.rb` or `features/modified_execution_paths/factory_boty_strategy/*_spec.rb`

**Review by pluff — CHANGES_REQUESTED** [review]

File: features/evaluation_spec.rb (hunk 10 above)
Comment (jaimerson) [reply features/evaluation_spec.rb]: I don't see much value in it, especially with the added complexity of handling overriding in nested contexts.

Comment (pluff) [reply features/evaluation_spec.rb]: there is no complexity. its literally more explicit alias.
```
def map_generator_config(&block)
  let(:map_generator_config, &block)
end
```

**Review by pluff — APPROVED** [review]

**Review by EPecherkin — APPROVED** [review]

**Pull request merged**